flocker/control/test/test_protocol.py

```diff
@@ -111,16 +115,76 @@ def test_deployment(self):
+def build_control_amp_service(test):
     """
-    Tests for ``ControlServiceLocator``.
+    Create a new ``ControlAMPService``.
+
+    :param TestCase test: The test this service is for.
+
+    :return ControlAMPService: Not started.
+    """
+    cluster_state = ClusterStateService()
+    cluster_state.startService()
+    test.addCleanup(cluster_state.stopService)
+    persistence_service = ConfigurationPersistenceService(
+        None, FilePath(test.mktemp()))
+    persistence_service.startService()
+    test.addCleanup(persistence_service.stopService)
+    return ControlAMPService(cluster_state, persistence_service,
+                             TCP4ServerEndpoint(MemoryReactor(), 1234))
+
+
+class ControlAMPTests(SynchronousTestCase):
+    """
+    Tests for ``ControlAMP`` and ``ControlServiceLocator``.
     """
     def setUp(self):
-        self.cluster_state = ClusterStateService()
-        self.cluster_state.startService()
-        self.addCleanup(self.cluster_state.stopService)
-        self.client = LoopbackAMPClient(ControlServiceLocator(
-            self.cluster_state))
+        self.control_amp_service = build_control_amp_service(self)
+        self.protocol = ControlAMP(self.control_amp_service)
+        self.client = LoopbackAMPClient(self.protocol.locator)
+
+    def test_connection_made(self):
+        """
+        When a connection is made the ``ControlAMP`` is added to the services
+        set of connections.
+        """
+        marker = object()
+        self.control_amp_service.connections.add(marker)
+        current = self.control_amp_service.connections.copy()
+        self.protocol.makeConnection(StringTransport())
+        self.assertEqual((current, self.control_amp_service.connections),
+                         ({marker}, {marker, self.protocol}))
+
+    def test_connection_made_send_cluster_status(self):
+        """
+        When a connection is made the cluster status is sent to the new client.
+        """
+        sent = []
+        self.patch(self.protocol, "callRemote",
+                   lambda *args, **kwargs: sent.append((args, kwargs)))
+        self.control_amp_service.configuration_service.save(TEST_DEPLOYMENT)
+        self.control_amp_service.cluster_state.update_node_state(
+            u"example3", NODE_STATE)
+
+        self.protocol.makeConnection(StringTransport())
+        cluster_state = self.control_amp_service.cluster_state.as_deployment()
+        self.assertEqual(
+            sent[0],
+            (((ClusterStatusCommand,),
+              dict(configuration=TEST_DEPLOYMENT,
+                   state=cluster_state))))
+
+    def test_connection_lost(self):
+        """
+        When a connection is lost the ``ControlAMP`` is removed the services
+        set of connections.
+        """
+        marker = object()
+        self.control_amp_service.connections.add(marker)
+        self.patch(self.protocol, "callRemote", lambda *args, **kwargs: None)
```

Comment:
``callRemote`` should return a Deferred. Does it matter that the patch function doesn't?